Convert SDK codebase to TypeScript

nan-li opened the pull request:
# Description

## One Line Summary

Convert the SDK codebase to TypeScript.

## Summary of Public Interface Changes
The original changes in this PR lived in the grey area of whether those constituted major breaking changes. The reason for this is that some of existing public interface and types were incorrect, and if we were to correct them, users may encounter compiler errors. 

The decision was reached to **not** create any changes or errors for someone using JavaScript, who may have discovered these to be incorrect, and is not necessarily following the public interface. As a result, some of the previously incorrect public interface and types are now corrected, and users on TypeScript will now encounter compiler errors. This is preferable as it can be considered a bug fix, and will not quietly fail for a JavaScript user unaware of the changes.

**These changes are:**

1. The interface for `InAppMessageAction` → keys are renamed and 2 additional properties are added (`outcomes` and `tags`)
```typescript
// Before
interface InAppMessageAction {
    closes_message  : boolean;
    first_click     : boolean;
    click_name      ?: string;
    click_url       ?: string;
}

// After
interface InAppMessageAction {
    closesMessage  : boolean;
    firstClick     : boolean;
    clickName      ?: string;
    clickUrl       ?: string;
    outcomes       ?: object[];
    tags           ?: object;
}
```

2. The interface for `OSNotification` → `groupedNotifications` added to interface
```typescript
export interface OSNotification {
    ...
    // android only
    groupedNotifications    ?: object[]; // 👈🏼 added to interface
    ...
}
```

3. The interface for `OpenedEventAction` (which is part of the response to `setNotificationOpenedHandler`) → `actionId` added to interface
```typescript
// Before
interface OpenedEventAction {
    type : OpenedEventActionType
}

// After
interface OpenedEventAction {
    actionId   ?: string;
    type       : OpenedEventActionType;
}
```

4. For adding triggers, the value parameter has been expanded to include `number` and `boolean`, and narrowed down when adding multiple triggers at once. 
```typescript
addTrigger(key: string, value: string): void; // Before
addTrigger(key: string, value: string | number | boolean): void; // After

addTriggers(triggers: object): void; // Before
addTriggers(triggers: {[key: string]: string | number | boolean}): void; // After
```

5. For the 2 methods `requiresUserPrivacyConsent` and `isLocationShared`, iOS passes an `object` version of the response to users while Android will pass a `boolean` directly. The iOS-only method `registerForProvisionalAuthorization` passes an `object` as well while Android does not trigger the handler. To minimize breaking changes, let's keep the response the same, but _update the interface_ to reflect reality by using union types.
* Admittedly, this is a bit ugly so we should update these in the next **major** release.
```typescript
registerForProvisionalAuthorization(handler?: (response: boolean) => void): void; // Before
registerForProvisionalAuthorization(handler?: (response: { accepted: boolean }) => void): void; // After

requiresUserPrivacyConsent(handler: (response: boolean) => void): void; // Before
requiresUserPrivacyConsent(handler: (response: boolean | { value: boolean }) => void): void; // After

isLocationShared(handler: (response: boolean) => void): void; // Before
isLocationShared(handler: (response: boolean | { value: boolean }) => void): void; // After
```

6. The interface for `setSMSNumber`, `logoutSMSNumber` and `setEmail`, `logoutEmail` → updated to accept 2 callbacks for `onSuccess` and `onFailure`
```typescript
setSMSNumber(smsNumber: string, authCode?: string, handler?: Function): void; // Before
setSMSNumber(smsNumber: string, authCode?: string, onSuccess?: Function, onFailure?: Function): void; // After

logoutSMSNumber(onSuccess?: Function, onFailure?: Function): void; // Before
logoutSMSNumber(handler?: Function): void; // After
    
setEmail(email: string, authCode?: string, handler?: Function): void; // Before
setEmail(email: string, authCode?: string, onSuccess?: Function, onFailure?: Function): void; // After

logoutEmail(handler?: Function): void; // Before
logoutEmail(onSuccess?: Function, onFailure?: Function): void; // After
```

7. Corrects the interface for `postNotification` method from `string` of the payload to an `object`.
```typescript
// Before
postNotification(notificationObjectString: string, onSuccess?: (success: object) => void, onFailure?: (failure: object) => void): void;

// After
postNotification(notificationObject: object, onSuccess?: (success: object) => void, onFailure?: (failure: object) => void): void;
```

## Details

### Motivation

TypeScript can support writing code that is less likely to contain bugs and is easier to maintain. The changes in this PR also will fix some misc bugs.
- [x] https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/734
- [x] https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/757

I tried to make the changes in this PR digestible by splitting up into many commits. The diffs in GitHub weren't great and it was hard to see what actually changes from commit to commit. Most notably, the `OneSignalPlugin` file changes are split into many commits.

### Scope

No new methods are added or removed. Previously there were some inconsistencies with the types we declared and what actually happens. Because of this, this PR can be considered as a semi-"breaking change".

For example, the interface for `InAppMessageAction` previously was snake_case, but what we were actually passing to users' callbacks were in camelCase. This interface is now updated to camelCase. We also added the 2 missing properties for `outcomes` and `tags` to the interface.

```
    closes_message  : boolean;
    first_click     : boolean;
    click_name      ?: string;
    click_url       ?: string;
```

```
    closesMessage  : boolean;
    firstClick     : boolean;
    clickName      ?: string;
    clickUrl       ?: string;
    outcomes       ?: object[];
    tags           ?: object;
```

The functions for setting email and SMS had previously wrong interface. The previous interface declared a single handler callback to be passed, but the actual method should be passed 2 callbacks for `onSuccess` and `onFailure`
- `setSMSNumber`, `logoutSMSNumber` and `setEmail`, `logoutEmail`
```typescript
    setEmail(email: string, authCode?: string, handler?: Function): void; // 👈🏼 previous interface

    setEmail(email: string, authCode?: string, onSuccess?: Function, onFailure?: Function): void; // 👈🏼 correct interface
```

The `postNotification` method's public interface has been corrected from accepting a `string` of the payload to an `object`.

**NOT UPDATED**: The `OneSignalPushProxy.js` file for windows phone under `/src/windows` which has been effectively deprecated.

### OneSignalPlugin File Changes

- deleted `OneSignal.completeNotification` method as developers do not use this directly, and we don't use it in the SDK.
- deleted this since due to a bug, it wasn't even used, and we already have `PermissionStatus`:
```javascript
OneSignalPlugin.prototype.OSNotificationPermission = {
    NotDetermined: 0,
    Authorized: 1,
    Denied: 2
};
```

### Added Scripts

In `package.json`, we added `prepare` and `build` scripts. `prepare` is run before we publish the SDK and it calls the `build` script to compile the TypeScript code into JavaScript according to our `tsconfig.json` file. Users of the SDK will then rely on the generated `.js` and `d.ts` files that will be generated under a new folder called `dist`.

```json
  "scripts": {
    "prepare": "yarn run build",
    "build": "tsc"
  }
```

###  ‼️ TODO

For email and SMS subscription change, the keys received from the native SDK is `isSubscribed`. We then change this key name to `isEmailSubscribed` and `isSMSSubscribed` and pass along to the user. However, after converting to TypeScript, I couldn't figure out the right way to do this. In the react-native-onesignal SDK, this type is also also inaccurate.

So, there is still the task left of modifying the `isSubscribed` property to `isEmailSubscribed` and `isSMSSubscribed` on their subscription change events, which can be a followup PR.

The followup PR is here: https://github.com/OneSignal/OneSignal-Cordova-SDK/pull/801

# Testing

## Manual testing

Tested with a minimal Ionic Capacitor Angular application using TypeScript.
- Added with `npm install onesignal-cordova-plugin` using this branch.
- Pixel Android emulator API 30
- iPhone 13 physical device
- [x] Tested access of public models such as `OSNotification`, `OutcomeEvent`, `LogLevel`, etc.
- [x] Tested access and setting of all the handlers and observers, and checking the response received.
- [x] Tested getting and dismissing IAMs
- [x] Tested receiving notifications in foreground and background

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes 👈🏼 sort of because some public types were previously wrong
# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [ ] I have included test coverage for these changes, or explained why they are not needed
   - [ ] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798)
<!-- Reviewable:end -->

rgomezp commented on src/android/com/onesignal/cordova/OneSignalPush.java:
```diff
@@ -506,10 +507,23 @@ public CordovaInAppMessageClickHandler(CallbackContext inCallbackContext) {
+          actionJSON.put("closesMessage", resultJSON.getBoolean("closes_message"));
+        }
+        actionJSON.put("clickName", resultJSON.optString("click_name", null));
+        actionJSON.put("clickUrl", resultJSON.optString("click_url", null));
```
Why do you say the types are incorrect? If you are converting to camelCase, then it is snake case currently which is consistent with RN. We have used snake case for quite a while.

We can convert to camel case, but this is a breaking change so we would need to make a major release.

rgomezp commented on www/index.ts:
```diff
@@ -225,7 +226,7 @@ registerForProvisionalAuthorization(handler?: (response: boolean) => void): void
 // Only applies to iOS (does nothing on Android as it always silently registers without user permission)
 promptForPushNotificationsWithUserResponse(handler?: (response: boolean) => void): void {
-    const internalCallback = (data) => {
+    const internalCallback = (data: any) => {
```
make a new type. avoid using any

nan-li replied:
I added a commit to pass the boolean from the iOS bridge so we don't have to extract. I did this for some other iOS bridge methods that corrected the callback received by the user. The commit has more details.

jkasten2 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798)* status: 0 of 24 files reviewed, 1 unresolved discussion (waiting on @emawby and @nan-li)

___
*[`package.json` line 12 at r2](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-N5r7WXb4vLtrNAXc6DM:-N5r7WXb4vLtrNAXc6DN:br4p0ru) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/c05ad8dcfce440ebea1d33d6978db42566f60418/package.json#L12)):*
> ```JSON
>   "scripts": {
>     "prepare": "npm run build",
>     "build": "tsc"
> ```

Does this introduce any new requirements? Do they need to install `tsc` or does that come with Cordova and Ionic?


<!-- Sent from Reviewable.io -->

nan-li commented on src/ios/OneSignalPush.m:
```diff
@@ -319,7 +319,7 @@ - (void)disablePush:(CDVInvokedUrlCommand*)command {
     postNotificationCallbackId = command.callbackId;
 
-    [OneSignal postNotification:command.arguments[0]
+    [OneSignal postNotificationWithJsonString:command.arguments[0]
```
This is a good call. We do need to revisit this because currently, although the public interface is a string, it would not have worked if a user passes a string, there would be errors.

nan-li replied:
Added commit to update interface to `object`

nan-li commented on package.json:
```diff
@@ -5,8 +5,12 @@
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "prepare": "npm run build",
```
Maybe use `yarn` instead. And commit a lock file?

jkasten2 reviewed (changes requested):
> Reviewed 1 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798)* status: 1 of 24 files reviewed, 1 unresolved discussion (waiting on @emawby and @nan-li)

___
*[`package.json` line 12 at r2](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-N5r7WXb4vLtrNAXc6DM:-N5rCq3m0AsuLGgQUJu_:bgns6gy) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/c05ad8dcfce440ebea1d33d6978db42566f60418/package.json#L12)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Does this introduce any new requirements? Do they need to install `tsc` or does that come with Cordova and Ionic?
</blockquote></details>

Or is the run when we ship an release? Therefore the ts.d file and .js files are what is actually published?


<!-- Sent from Reviewable.io -->

nan-li reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798)* status: 1 of 24 files reviewed, 1 unresolved discussion (waiting on @emawby and @jkasten2)

___
*[`package.json` line 12 at r2](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-N5r7WXb4vLtrNAXc6DM:-N5rCopqCY7Lfj5EPuTI:bom8upl) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/c05ad8dcfce440ebea1d33d6978db42566f60418/package.json#L12)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Or is the run when we ship an release? Therefore the ts.d file and .js files are what is actually published?
</blockquote></details>

Yup, this is run before we publish and users will rely on the generated `.js` and `d.ts` files that will be generated under a new folder called `dist`


<!-- Sent from Reviewable.io -->

nan-li commented on www/Subscription.ts:
```diff
@@ -0,0 +1,73 @@
+        this.emailAddress = deviceState.emailAddress;
+        this.smsUserId = deviceState.smsUserId;
+        this.smsNumber = deviceState.smsNumber;
+        // rename the subscrbied properties to align with existing type definition
```
fixed

nan-li commented on www/index.ts:
```diff
@@ -25,31 +25,39 @@
+    private _smsSubscriptionObserverList: ((event:ChangeEvent<SMSSubscriptionChange>)=>void)[] = [];
+
+    private _processFunctionList<ObserverChangeEvent>(array: ((event:ChangeEvent<ObserverChangeEvent>)=>void)[], param: ChangeEvent<ObserverChangeEvent>): void {
+        for (let i = 0; i < array.length; i++)
```
Fixed brackets in nit commit

nan-li commented:
Summarizing the changes I added since last review are:
- Updated package.json to use `yarn` instead of `npm`. The lockfile generated is actually empty because the SDK names no dependencies, so no lockfile is committed.
- Add `outcomes` and `tags` to `InAppMessageAction` and tested this via the IAM click handler on iPhone and Android when these are present and when these are not.
- Update how iOS passes booleans to the JavaScript code and users' callbacks. Now, several methods will pass the boolean directly instead of as an object such as `{accepted: true}`. Android has always been passing the boolean, so now both platforms are consistent. This also fixes type inconsistencies as the public interface declares that users will receive a boolean as a callback response, but we had previously been passing them an object on iOS and a boolean on Android.
- Correct the `postNotification` interface to take `object` instead of `string` (which was erroneous).
- A nit commit to remove some trailing white space, spaces, add brackets to JavaScript code, and add two comments.

jkasten2 reviewed (changes requested):
> Reviewed 4 of 20 files at r1, 3 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798)* status: 7 of 24 files reviewed, 3 unresolved discussions (waiting on @emawby, @jkasten2, @nan-li, and @rgomezp)

___
*[`package.json` line 12 at r2](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-N5r7WXb4vLtrNAXc6DM:-N6FVegGCXQ1sRxWbFEi:b-ksemp2) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/c05ad8dcfce440ebea1d33d6978db42566f60418/package.json#L12)):*
<details><summary><i>Previously, nan-li (Nan) wrote…</i></summary><blockquote>

Yup, this is run before we publish and users will rely on the generated `.js` and `d.ts` files that will be generated under a new folder called `dist`
</blockquote></details>

ah ok, would be good to include this in the PR details.
___
*[`www/index.ts` line 94 at r3](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-N6FXE8REKzzYmkOdWMC:-N6FXE8REKzzYmkOdWMD:bqzhbvn) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/48994c9718641a0c53c904de4c8bb210c6071392/www/index.ts#L94)):*
> ```XML
>         const foregroundParsingHandler = (notificationReceived: OSNotification) => {
> 
>             console.log("foregroundParsingHandler " + JSON.stringify(notificationReceived));
> ```

Log entry should removed or should start with OneSignal so the source of it is known.
___
*[`www/index.ts` line 587 at r3](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-N6FZVVE9Nt9YUrG58zo:-N6FZVVE9Nt9YUrG58zp:bz1nuf8) ([raw file](https://github.com/onesignal/onesignal-cordova-sdk/blob/48994c9718641a0c53c904de4c8bb210c6071392/www/index.ts#L587)):*
> ```XML
>         }
> 
>         var passToNativeParams = [externalId];
> ```

Could be `const`


<!-- Sent from Reviewable.io -->

rgomezp commented on package.json:
```diff
@@ -5,8 +5,12 @@
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "prepare": "yarn run build",
+    "build": "tsc"
```
Not a blocker, but would be nice to add ESLint as a prebuild step:

`"prepare": "yarn run lint && yarn run build"`
`"lint": "eslint . --ext .ts"`

nan-li replied:
Good idea, let's follow up on this later since we need to make the configuration, and it can also be applied to `react-native-onesignal` which doesn't have a linter also.

jkasten2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 20 files at r1, 5 of 12 files at r2, 5 of 5 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-:-N6njL-d2QhpaKRP1rds:bnfp4nl)* status: 23 of 24 files reviewed, all discussions resolved (waiting on @emawby and @jkasten2)



<!-- Sent from Reviewable.io -->

nan-li commented:
Added 2 commits to the PR. 
- The first caught 2 issues while I was combing through the PR changes. 
- The second reverts https://github.com/OneSignal/OneSignal-Cordova-SDK/pull/798/commits/3eb8ad71d499dcf4e318e63d2f1f8cf003197b04 and keeps iOS boolean callbacks the same, and updates the interface.

I also added a section called **Summary of Public Interface Changes** to the PR description that lists all the public-facing changes of this PR, making it easier to see at a glance everything that changed.

jkasten2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r5, 1 of 2 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-cordova-sdk/798#-:-N7cOqUZ60nFMMgCol0P:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @emawby and @rgomezp)



<!-- Sent from Reviewable.io -->
